modules/phase_field/src/materials/ComputeGBMisorientationType.C

```diff
@@ -0,0 +1,153 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "ComputeGBMisorientationType.h"
+#include "SolutionUserObject.h"
+#include <fstream>
+
+registerMooseObject("PhaseFieldApp", ComputeGBMisorientationType);
+
+InputParameters
+ComputeGBMisorientationType::validParams()
+{
+  InputParameters params = Material::validParams();
+  params.addClassDescription("Calculate types of grain boundaries in a polycrystalline sample");
+  params.addRequiredParam<UserObjectName>("grain_tracker",
+                                          "The GrainTracker UserObject to get values from.");
+  params.addRequiredParam<UserObjectName>("ebsd_reader", "The EBSDReader GeneralUserObject");
+  params.addRequiredCoupledVarWithAutoBuild(
+      "v", "var_name_base", "op_num", "Array of coupled variables");
+  params.addRequiredParam<FileName>("file_name", "Misorientation angle data file name");
+  params.addParam<Real>("angle_threshold", 15, "Max LAGB Misorientation angle");
+  return params;
+}
+
+ComputeGBMisorientationType::ComputeGBMisorientationType(const InputParameters & parameters)
+  : Material(parameters),
+    _grain_tracker(getUserObject<GrainTracker>("grain_tracker")),
+    _ebsd_reader(getUserObject<EBSDReader>("ebsd_reader")),
+    _file_name(getParam<FileName>("file_name")),
+    _op_num(coupledComponents("v")),
+    _vals(coupledValues("v")),
+    _angle_threshold(getParam<Real>("angle_threshold")),
+    _gb_type(declareADProperty<Real>("gb_type"))
+{
+  readFile();
+}
+
+void
+ComputeGBMisorientationType::computeQpProperties()
+{
+
+  // Find out the number of boundary unique_id and save them
+  std::vector<unsigned int> gb_pairs;
+  std::vector<Real> gb_op_pairs;
+  MooseIndex(_vals) val_index = -1;
+
+  const auto & op_to_grains = _grain_tracker.getVarToFeatureVector(_current_elem->id());
+  for (auto & unique_id : op_to_grains)
+  {
+    val_index += 1;
+    if (unique_id == FeatureFloodCount::invalid_id)
+      continue;
+
+    auto grain_id = _ebsd_reader.getFeatureID(unique_id);
+
+    gb_pairs.push_back(grain_id);
+    gb_op_pairs.push_back((*_vals[val_index])[_qp]);
+  }
+
+  // Compute GB type by the number of id
+  switch (gb_pairs.size())
+  {
+    case 0:
+    {
+      break;
```

Comment:
better set `_gb_type[_qp]` to something well defined, otherwise it will keep the value from the previous QP/element.